File: backend/src/main/java/com/nforce/onehr/validation/PasswordPolicy.java
package com.nforce.onehr.validation;

/**
 * Single source of truth for the platform's password policy. Referenced by
 * {@link ValidPassword}/{@link PasswordConstraintValidator} (Bean Validation, enforced on
 * every password-setting API request) so the rules can't drift between endpoints or between
 * a DTO annotation and hand-written service code. If the policy ever changes, change it here
 * only — every flow that sets a password (initial setup, change, and any future reset flow)
 * picks it up automatically via {@code @ValidPassword}.
 */
public final class PasswordPolicy {

    private PasswordPolicy() {
    }

    public static final int MIN_LENGTH = 8;
    public static final int MAX_LENGTH = 128;
    public static final int MIN_CHARACTER_CLASSES = 3;

    public static final String LENGTH_MESSAGE =
            "Password must be between " + MIN_LENGTH + " and " + MAX_LENGTH + " characters";
    public static final String SPACE_MESSAGE = "Password cannot contain spaces.";
    public static final String COMPLEXITY_MESSAGE =
            "Password must include at least " + MIN_CHARACTER_CLASSES
                    + " of: uppercase letter, lowercase letter, number, special character";

    /** True if the password contains a space anywhere — leading, trailing, or between characters. */
    public static boolean containsSpace(String password) {
        return password.indexOf(' ') >= 0;
    }

    /** Character-class score used by both the complexity check and the frontend strength meter. */
    public static int characterClassScore(String password) {
        int score = 0;
        if (password.chars().anyMatch(Character::isUpperCase)) score++;
        if (password.chars().anyMatch(Character::isLowerCase)) score++;
        if (password.chars().anyMatch(Character::isDigit)) score++;
        if (password.matches(".*[!@#$%^&*()_+\\-=\\[\\]{};':\"\\\\|,.<>/?].*")) score++;
        return score;
    }
}

File: frontend/src/utils/passwordPolicy.ts
// Single source of truth for the frontend's password policy. Mirrors the backend's
// PasswordPolicy (backend/src/main/java/com/nforce/onehr/validation/PasswordPolicy.java) so
// the client can give instant feedback, but the backend re-enforces every rule independently —
// this file only improves UX, it is never the security boundary.

export const PASSWORD_MIN_LENGTH = 8;
export const PASSWORD_MAX_LENGTH = 128;
export const PASSWORD_MIN_CHARACTER_CLASSES = 3;
export const PASSWORD_SPACE_MESSAGE = 'Password cannot contain spaces.';

/** True if the password contains a space anywhere — leading, trailing, or between characters. */
export function containsSpace(password: string): boolean {
  return password.includes(' ');
}

export const PASSWORD_CRITERIA = [
  { label: 'Uppercase letter', test: (p: string) => /[A-Z]/.test(p) },
  { label: 'Lowercase letter', test: (p: string) => /[a-z]/.test(p) },
  { label: 'Number', test: (p: string) => /[0-9]/.test(p) },
  { label: 'Special character', test: (p: string) => /[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>/?]/.test(p) },
];

export function passwordCharacterClassScore(password: string): number {
  return PASSWORD_CRITERIA.filter((c) => c.test(password)).length;
}

/** Drives the 4-bar strength meter; caps at "fair" until the length floor is met. */
export function passwordStrengthScore(password: string): number {
  if (!password) return 0;
  const classScore = passwordCharacterClassScore(password);
  return password.length < PASSWORD_MIN_LENGTH ? Math.min(classScore, 2) : classScore;
}

/**
 * Validates a new password against the full policy. Returns the first failing rule's
 * message, or null if the password is acceptable. Order matters: length and spaces are
 * checked before complexity so the user always sees the most fundamental problem first.
 */
export function validateNewPassword(password: string): string | null {
  if (!password) return 'New password is required';
  if (containsSpace(password)) {
    return PASSWORD_SPACE_MESSAGE;
  }
  if (password.length < PASSWORD_MIN_LENGTH) {
    return `Must be at least ${PASSWORD_MIN_LENGTH} characters`;
  }
  if (password.length > PASSWORD_MAX_LENGTH) {
    return `Must be no more than ${PASSWORD_MAX_LENGTH} characters`;
  }
  if (passwordCharacterClassScore(password) < PASSWORD_MIN_CHARACTER_CLASSES) {
    return `Must include at least ${PASSWORD_MIN_CHARACTER_CLASSES} of: uppercase, lowercase, number, special character`;
  }
  return null;
}
